pkg/cueloader/load.go

```diff
@@ -0,0 +1,197 @@
+		l.LoadEnd = time.Now()
+		fmt.Println("LoadEnd")
+		if err := instance.Err; err != nil {
+			l.Err = errors.Join(l.Err, instance.Err)
+			return l
+		}
+		if l.Err != nil {
+			return l
+		}
+
+		if mu != nil {
+			mu.Lock()
+			defer mu.Unlock()
+		}
+		fmt.Println("BuildStart")
+		l.BuildStart = time.Now()
+		l.BuildValue = cc.BuildInstance(instance)
+		l.BuildEnd = time.Now()
+		fmt.Println("BuildEnd")
+
+		if l.BuildValue.Err() != nil {
+			fmt.Println("BuildErr")
+			l.Err = l.BuildValue.Err()
+			return l
+		}
+
+		l.Value = l.BuildValue
+		fmt.Println("TransformStart")
+		l.TransformStart = time.Now()
+		for _, transform := range transforms {
+			// fmt.Println("->", l.Value)
+			l.Value = transform(l.Value)
+			if l.Value.Err() != nil {
+				fmt.Println("TransformErr")
+				l.Err = l.Value.Err()
+				return l
+			}
+		}
+		// fmt.Println("->", l.Value)
+		l.TransformEnd = time.Now()
+		fmt.Println("TransformEnd")
+
+		return l
+	}
+}
+
+func CopyConfigAndReadFilesIntoOverrides(config *load.Config) (map[string]string, *load.Config, error) {
+	copy := *config
+	files := map[string]string{}
+	err := AddCueFiles(files, copy.Dir, "", false)
+	if err != nil {
+		return files, &copy, err
+	}
+	// Reset the overlay
+	overlay := map[string]load.Source{}
+	for k, v := range files {
+		overlay[k] = load.FromString(v)
+	}
+	copy.Overlay = overlay
+	copy.Tags = append([]string{}, copy.Tags...)
+	return files, &copy, err
+}
+
+func NewCueConfigWatcher(ctx context.Context, config *load.Config, cb func(err error, files map[string]string, config *load.Config)) error {
+	watcher, err := NewWatcher(ctx, func(err error) {
+		if err != nil {
+			cb(err, nil, nil)
+			return
+		}
+		files, copy, err := CopyConfigAndReadFilesIntoOverrides(config)
+		cb(err, files, copy)
+	})
+	if err != nil {
+		return err
+	}
+	watcher.Add(config.Dir)
+
+	return nil
+}
+
+
+func NewCueConfigWatcherFromURL(ctx context.Context, url string, cb func (err error, files map[string]string, config *load.Config)) error {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return err
+	}
+	log.Printf("making request of %s", url)
+	return ReadStreamEvents(http.DefaultClient, req, func (event *Event) error {
+		files, config, err := Unmarshal(event.Data)
+		log.Printf("%d files dir=%s tags=%#v", len(files), config.Dir, config.Tags)
+		cb(err, files, config)
+		return nil
+	})
+}
+
+func NewWatcher(ctx context.Context, cb func(err error)) (*fsnotify.Watcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		fmt.Printf("WARN: error starting fsnotify, live editing disabled!\n")
+		return nil, err
+	}
+
+	debounceInterval := 100 * time.Millisecond
+	go func() {
+		defer log.Println("done watching")
+		debounce := time.NewTimer(debounceInterval)
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				log.Println("event: ", event)
+				debounce.Reset(debounceInterval)
```

Comment:
Timers have some annoying caveats around how to reset them:

> For a Timer created with NewTimer, Reset should be invoked only on stopped or expired timers with drained channels.
> If a program has already received a value from t.C, the timer is known to have expired and the channel drained, so t.Reset can be used directly. If a program has not yet received a value from t.C, however, the timer must be stopped and—if Stop reports that the timer expired before being stopped—the channel explicitly drained:
> ```
> if !t.Stop() {
> 	<-t.C
> }
> t.Reset(d)
> ```
> This should not be done concurrent to other receives from the Timer's channel.

https://pkg.go.dev/time@go1.22.0#Timer.Reset

Reply:
Thanks for calling out the Reset spec. It seems that we should call `.Stop()` and then only `.Reset(...)` if returns false. Inside of our loop we can check if we slept for long enough and if not then call reset on our own.

It feels like this might still drop an event in some circumstances but I haven't figured out how yet.